File: 00-code(源代码)/src/com/hi/dhl/algorithms/offer/_22/java/Solution.java
package com.hi.dhl.algorithms.offer._22.java;

import com.hi.dhl.algorithms.model.ListNode;

/**
 * <pre>
 *     author: dhl
 *     date  : 2020/8/29
 *     desc  :
 * </pre>
 */
class Solution {
    public ListNode getKthFromEnd(ListNode head, int k) {
        if (head == null || k < 0) return head;

        ListNode fast = head;
        for (int i = 1; i < k; i++) {
            fast = fast.next;
        }

        ListNode slow = head;
        while (fast.next != null) {
            fast = fast.next;
            slow = slow.next;
        }
        return slow;
    }
}

File: 00-code(源代码)/src/com/hi/dhl/algorithms/offer/_22/kotlin/Solution.kt
package com.hi.dhl.algorithms.offer._22.kotlin

import com.hi.dhl.algorithms.model.ListNode

/**
 * <pre>
 *     author: dhl
 *     date  : 2020/8/29
 *     desc  :
 * </pre>
 */
class Solution {
    fun getKthFromEnd(head: ListNode?, k: Int): ListNode? {
        if (head == null || k < 0) return head
        var fast = head;

        (1 until k).forEach {
            fast = fast?.next
        }

        var slow = head;
        while (fast?.next != null) {
            fast = fast?.next;
            slow = slow?.next;
        }

        return slow;
    }
}